File: graph_components/dfs/src/DfsDemo.py
# Python
# グラフの連結成分: DFS

class GraphData:
    def __init__(self):
        # 隣接ノードとその辺の重みを格納します。
        # キーは頂点、値はその頂点に隣接する頂点と重みの辞書です。
        # 例: { 'A': {'B': 1, 'C': 4}, 'B': {'A': 1, 'C': 2, 'D': 5}, ... }
        self._data = {}

    def get(self):
        # グラフの内部データを取得します。
        return self._data

    def get_vertices(self):
        # グラフの全頂点をリストとして返します。
        return list(self._data.keys())
    
    def add_vertex(self, vertex):
        # 新しい頂点をグラフに追加します。
        if vertex not in self._data:
            self._data[vertex] = []
            return True
        # 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return True
    
    def add_edge(self, vertex1, vertex2, weight):
        # 両頂点間に辺を追加します。重みを指定します。
        # 頂点がグラフに存在しない場合は追加します。
        if vertex1 not in self._data:
            self.add_vertex(vertex1)
        if vertex2 not in self._data:
            self.add_vertex(vertex2)
        
        # 両方向に辺を追加する（無向グラフ）
        # 既に同じ辺が存在する場合は重みを更新するかどうか？
        # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
        # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
        
        # vertex1 -> vertex2 の辺を追加（重み付き）
        edge_exists_v1v2 = False
        for i, (neighbor, _) in enumerate(self._data[vertex1]):
            if neighbor == vertex2:
                self._data[vertex1][i] = (vertex2, weight) # 既に存在する場合は重みを更新
                edge_exists_v1v2 = True
                break
        if not edge_exists_v1v2:
            self._data[vertex1].append((vertex2, weight))

        # vertex2 -> vertex1 の辺を追加（重み付き）
        edge_exists_v2v1 = False
        for i, (neighbor, _) in enumerate(self._data[vertex2]):
            if neighbor == vertex1:
                self._data[vertex2][i] = (vertex1, weight) # 既に存在する場合は重みを更新
                edge_exists_v2v1 = True
                break
        if not edge_exists_v2v1:
            self._data[vertex2].append((vertex1, weight))
        
        return True
    
    def clear(self):
        # グラフを空にする
        self._data = {}
        return True

    def _dfs(self, vertex, visited, current_component):
        # 現在の頂点を訪問済みにマークし、現在の成分に追加
        visited.add(vertex)
        current_component.append(vertex)

        # 隣接する頂点を探索
        # self._data.get(vertex, []) は、vertexが存在しない場合でも
        # エラーにならず空リストを返す安全な方法
        # ここで取得されるのは [(neighbor, weight), ...] のリストです。
        for neighbor_info in self._data.get(vertex, []):
            # neighbor_info は (neighbor, weight) のタプルなので、
            # タプルの0番目の要素（隣接頂点そのもの）を取得します。
            neighbor_vertex = neighbor_info[0]

            # まだ訪問していない隣接頂点に対して再帰的にDFSを呼び出す
            if neighbor_vertex not in visited:
                self._dfs(neighbor_vertex, visited, current_component)

    def get_connected_components(self):
        visited = set() # 訪問済み頂点を記録するセット
        connected_components = [] # 連結成分を格納するリスト

        # グラフの全頂点を順にチェック
        for vertex in self.get_vertices():
            # まだ訪問していない頂点からDFSを開始
            if vertex not in visited:
                current_component = [] # 現在の連結成分を格納するリスト
                # DFSヘルパー関数を呼び出し、現在の連結成分を探索
                self._dfs(vertex, visited, current_component)
                # 探索で見つかった連結成分を結果リストに追加
                connected_components.append(current_component)

        return connected_components

def main():
    print("Dfs TEST -----> start")

    print("\nnew")
    graph_data = GraphData()
    print(f"  現在のデータ: {graph_data.get()}")

    print("\nadd_edge")
    graph_data.clear()
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('B', 'D', 2), ('D', 'A', 1), ('A', 'C', 2), ('B', 'D', 2)]
    for input in inputList:
        print(f"  入力値: {input}")
        output = graph_data.add_edge(*input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")
    print("\nget_connected_components")
    output = graph_data.get_connected_components()
    print(f"  連結成分: {output}")

    print("\nadd_edge")
    graph_data.clear()
    inputList = [('A', 'B', 4), ('C', 'D', 4), ('E', 'F', 1), ('F', 'G', 1)]
    for input in inputList:
        print(f"  入力値: {input}")
        output = graph_data.add_edge(*input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")
    print("\nget_connected_components")
    output = graph_data.get_connected_components()
    print(f"  連結成分: {output}")

    print("\nadd_edge")
    graph_data.clear()        
    inputList = [('A', 'B', 4), ('B', 'C', 3), ('D', 'E', 5)]
    for input in inputList:
        print(f"  入力値: {input}")
        output = graph_data.add_edge(*input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")
    print("\nget_connected_components")
    output = graph_data.get_connected_components()
    print(f"  連結成分: {output}")

    print("\nadd_edge")
    graph_data.clear()
    inputList = []
    for input in inputList:
        print(f"  入力値: {input}")
        output = graph_data.add_edge(*input)
        print(f"  出力値: {output}")
    print(f"  現在のデータ: {graph_data.get()}")
    print("\nget_connected_components")
    output = graph_data.get_connected_components()
    print(f"  連結成分: {output}")

    print("Dfs TEST <----- end")

if __name__ == "__main__":
    main()

File: graph_components/dfs/src/DfsDemo.cpp
// C++
// グラフの連結成分: DFS

#include <iostream>
#include <unordered_map>
#include <vector>
#include <set>
#include <algorithm>
#include <stdexcept>

class GraphData {
private:
    // キーは頂点、値は隣接頂点と重みのペアのベクタ
    std::unordered_map<std::string, std::vector<std::pair<std::string, int>>> _data;

    void _dfs(const std::string& vertex, std::set<std::string>& visited, std::vector<std::string>& current_component) {
        visited.insert(vertex);
        current_component.push_back(vertex);

        // 隣接頂点を探索
        for (const auto& neighbor_info : _data[vertex]) {
            const std::string& neighbor_vertex = neighbor_info.first;
            if (visited.find(neighbor_vertex) == visited.end()) {
                _dfs(neighbor_vertex, visited, current_component);
            }
        }
    }

public:
    std::unordered_map<std::string, std::vector<std::pair<std::string, int>>> get() const {
        return _data;
    }

    std::vector<std::string> get_vertices() const {
        std::vector<std::string> vertices;
        for (const auto& pair : _data) {
            vertices.push_back(pair.first);
        }
        return vertices;
    }

    bool add_vertex(const std::string& vertex) {
        if (_data.find(vertex) == _data.end()) {
            _data[vertex] = {};
        }
        return true;
    }

    bool add_edge(const std::string& vertex1, const std::string& vertex2, int weight) {
        // 頂点が存在しない場合は追加
        add_vertex(vertex1);
        add_vertex(vertex2);

        // vertex1 -> vertex2の辺を追加
        bool edge_exists_v1v2 = false;
        for (auto& neighbor : _data[vertex1]) {
            if (neighbor.first == vertex2) {
                neighbor.second = weight;
                edge_exists_v1v2 = true;
                break;
            }
        }
        if (!edge_exists_v1v2) {
            _data[vertex1].emplace_back(vertex2, weight);
        }

        // vertex2 -> vertex1の辺を追加
        bool edge_exists_v2v1 = false;
        for (auto& neighbor : _data[vertex2]) {
            if (neighbor.first == vertex1) {
                neighbor.second = weight;
                edge_exists_v2v1 = true;
                break;
            }
        }
        if (!edge_exists_v2v1) {
            _data[vertex2].emplace_back(vertex1, weight);
        }

        return true;
    }

    bool clear() {
        _data.clear();
        return true;
    }

    std::vector<std::vector<std::string>> get_connected_components() {
        std::set<std::string> visited;
        std::vector<std::vector<std::string>> connected_components;

        for (const auto& vertex : get_vertices()) {
            if (visited.find(vertex) == visited.end()) {
                std::vector<std::string> current_component;
                _dfs(vertex, visited, current_component);
                connected_components.push_back(current_component);
            }
        }

        return connected_components;
    }
};

void print_graph_data(const GraphData& graph_data) {
    std::cout << "現在のデータ: {";
    for (const auto& vertex : graph_data.get()) {
        std::cout << vertex.first << ": [";
        for (const auto& neighbor : vertex.second) {
            std::cout << "(" << neighbor.first << ", " << neighbor.second << ") ";
        }
        std::cout << "], ";
    }
    std::cout << "}" << std::endl;
}

int main() {
    std::cout << "Dfs TEST -----> start" << std::endl;

    std::cout << "\nnew" << std::endl;
    GraphData graph_data;
    print_graph_data(graph_data);

    std::cout << "\nadd_edge" << std::endl;
    graph_data.clear();
    std::vector<std::tuple<std::string, std::string, int>> inputList = {
        {"A", "B", 4}, {"B", "C", 3}, {"B", "D", 2}, 
        {"D", "A", 1}, {"A", "C", 2}, {"B", "D", 2}
    };
    for (const auto& input : inputList) {
        std::cout << "  入力値: (" 
                  << std::get<0>(input) << ", " 
                  << std::get<1>(input) << ", " 
                  << std::get<2>(input) << ")" << std::endl;
        bool output = graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
        std::cout << "  出力値: " << std::boolalpha << output << std::endl;
    }
    print_graph_data(graph_data);
    std::cout << "\nget_connected_components" << std::endl;
    auto output = graph_data.get_connected_components();
    std::cout << "  連結成分: ";
    for (const auto& component : output) {
        std::cout << "[ ";
        for (const auto& vertex : component) {
            std::cout << vertex << " ";
        }
        std::cout << "] ";
    }
    std::cout << std::endl;

    std::cout << "\nadd_edge" << std::endl;
    graph_data.clear();
    inputList = {{"A", "B", 4}, {"C", "D", 4}, {"E", "F", 1}, {"F", "G", 1}};
    for (const auto& input : inputList) {
        std::cout << "  入力値: (" 
                  << std::get<0>(input) << ", " 
                  << std::get<1>(input) << ", " 
                  << std::get<2>(input) << ")" << std::endl;
        bool output = graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
        std::cout << "  出力値: " << std::boolalpha << output << std::endl;
    }
    print_graph_data(graph_data);
    std::cout << "\nget_connected_components" << std::endl;
    output = graph_data.get_connected_components();
    std::cout << "  連結成分: ";
    for (const auto& component : output) {
        std::cout << "[ ";
        for (const auto& vertex : component) {
            std::cout << vertex << " ";
        }
        std::cout << "] ";
    }
    std::cout << std::endl;

    std::cout << "\nadd_edge" << std::endl;
    graph_data.clear();
    inputList = {{"A", "B", 4}, {"B", "C", 3}, {"D", "E", 5}};
    for (const auto& input : inputList) {
        std::cout << "  入力値: (" 
                  << std::get<0>(input) << ", " 
                  << std::get<1>(input) << ", " 
                  << std::get<2>(input) << ")" << std::endl;
        bool output = graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
        std::cout << "  出力値: " << std::boolalpha << output << std::endl;
    }
    print_graph_data(graph_data);
    std::cout << "\nget_connected_components" << std::endl;
    output = graph_data.get_connected_components();
    std::cout << "  連結成分: ";
    for (const auto& component : output) {
        std::cout << "[ ";
        for (const auto& vertex : component) {
            std::cout << vertex << " ";
        }
        std::cout << "] ";
    }
    std::cout << std::endl;

    std::cout << "\nadd_edge" << std::endl;
    graph_data.clear();
    inputList = {};
    for (const auto& input : inputList) {
        std::cout << "  入力値: (" 
                  << std::get<0>(input) << ", " 
                  << std::get<1>(input) << ", " 
                  << std::get<2>(input) << ")" << std::endl;
        bool output = graph_data.add_edge(std::get<0>(input), std::get<1>(input), std::get<2>(input));
        std::cout << "  出力値: " << std::boolalpha << output << std::endl;
    }
    print_graph_data(graph_data);
    std::cout << "\nget_connected_components" << std::endl;
    output = graph_data.get_connected_components();
    std::cout << "  連結成分: ";
    for (const auto& component : output) {
        std::cout << "[ ";
        for (const auto& vertex : component) {
            std::cout << vertex << " ";
        }
        std::cout << "] ";
    }
    std::cout << std::endl;

    std::cout << "Dfs TEST <----- end" << std::endl;

    return 0;
}
